dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/NotificationSender.java

```diff
@@ -35,19 +35,83 @@
 import org.hisp.dhis.notification.logging.NotificationTriggerEvent;
 import org.hisp.dhis.notification.logging.NotificationValidationResult;
 import org.hisp.dhis.program.Enrollment;
+import org.hisp.dhis.program.Event;
+import org.hisp.dhis.program.notification.ProgramNotificationInstance;
+import org.hisp.dhis.program.notification.ProgramNotificationInstanceService;
+import org.hisp.dhis.program.notification.ProgramNotificationService;
 import org.hisp.dhis.program.notification.ProgramNotificationTemplate;
 import org.hisp.dhis.program.notification.ProgramNotificationTemplateService;
+import org.hisp.dhis.program.notification.template.snapshot.NotificationTemplateService;
 import org.hisp.dhis.programrule.api.NotificationEffect;
 import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
 
+/** Sends or schedule notification as a result of rule-engine evaluation. */
 @Slf4j
 @RequiredArgsConstructor
 @Component
-class NotificationHelper {
+public class NotificationSender {
+  public static final String LOG_MESSAGE = "Notification with id:%s has been scheduled";
+  private final ProgramNotificationInstanceService programNotificationInstanceService;
+  private final ProgramNotificationService programNotificationService;
+  private final NotificationTemplateService notificationTemplateService;
   private final ProgramNotificationTemplateService programNotificationTemplateService;
   private final NotificationLoggingService notificationLoggingService;
 
-  public void createLogEntry(ProgramNotificationTemplate template, Enrollment enrollment) {
+  @Transactional
+  public void send(NotificationEffect notificationEffect, Enrollment enrollment) {
+    ProgramNotificationTemplate template = getNotificationTemplate(notificationEffect);
+
+    NotificationValidationResult result = validate(template, enrollment);
+
+    if (result.isValid()) {
+      if (notificationEffect.date() != null) {
+        ProgramNotificationInstance notificationInstance =
+            notificationTemplateService.createNotificationInstance(
+                template, notificationEffect.date());
+        notificationInstance.setEvent(null);
+        notificationInstance.setEnrollment(enrollment);
+
+        programNotificationInstanceService.save(notificationInstance);
+
+        log.info(String.format(LOG_MESSAGE, template.getUid()));
```

Comment:
Logging should be done using parameterized messages https://www.slf4j.org/faq.html#logging_performance 

Should that be an info log? To me this is debug. We log to much IMHO. This is why we have the logging group combing through logs 😅 

Ask yourself who would want to see this and what should they do with it? Logs should be actionable by the operator of DHIS2.

Reply:
I just removed it.
I think it didn't really make any sense to log a scheduled notification and not the sent one.
Scheduled notifications are saved somewhere in the system. If there is any need to debug it will be better to check the DB or the endpoint if we have one to understand what happened